src/libsync/clientstatusreportingnetwork.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2023 by Oleksandr Zolotov <alex@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+#pragma once
+
+#include "owncloudlib.h"
+#include <common/result.h>
+#include "clientstatusreportingcommon.h"
+
+#include <QtCore/qglobal.h>
+#include <QtCore/qbytearray.h>
+#include <QtCore/qobject.h>
+#include <QtCore/qpointer.h>
+#include <QtCore/qsharedpointer.h>
+#include <QtCore/qstring.h>
+#include <QtCore/qtimer.h>
+
+namespace OCC {
+
+class Account;
+class ClientStatusReportingDatabase;
+struct ClientStatusReportingRecord;
+
+class OWNCLOUDSYNC_EXPORT ClientStatusReportingNetwork : public QObject
+{
+    Q_OBJECT
+public:
+    explicit ClientStatusReportingNetwork(Account *account, const QSharedPointer<ClientStatusReportingDatabase> database, QObject *parent = nullptr);
+    ~ClientStatusReportingNetwork() override;
+
+private:
+    void init();
+
+    [[nodiscard]] QVariantMap prepareReport() const;
+    void reportToServerSentSuccessfully();
+
+private slots:
+    void sendReportToServer();
+
+public:
+    [[nodiscard]] bool isInitialized() const;
+
+    static QByteArray classifyStatus(const ClientStatusReportingStatus status);
+
+    static int clientStatusReportingTrySendTimerInterval;
+    static quint64 repordSendIntervalMs;
+    // this must be set in unit tests on init
+    static QString dbPathForTesting;
+
+private:
+    QPointer<Account> _account = nullptr;
```

Comment:
why not `AccountPtr` ?
you even have an include just to get this type

Reply:
@mgallien
I do not have an include for this type. `accountfwd.h` was never included here.
 `ClientStatusReportingNetwork` must NOT have shared ownership for the `Account`, otherwise, the account will never get destructed as it will keep the ref counter at `1` and the `ClientStatusReportingNetwork` is destructed when the destructor of `Account` executes as the account keeps the pointer to `ClientStatusReporting` which in turn keeps the pointer to `ClientStatusReportingNetwork`. The shared pointer is obtained when the network job `JsonApiJob` is instantiated. Having a `QPointer` allows to have a valid `Account` during the `JsonApiJob` instantiation, and the `JsonApiJob` then destructs itself properly, hence decrementing the ref count for `Account`. So, to summarize, keeping an a shared pointer is not something that we need here. The `Account` owns the entire chain if classes for the reporting, so, it is assumed to be always valid, and `ClientStatusReportingNetwork` will get destroyed when it is not valid anymore.

More explanation:

- `Account` holds a unique pointer to `ClientStatusReporting`
- `ClientStatusReporting` keeps a unique pointer to `ClientStatusReportingNetwork`
- When `Account` gets destroyed, it calls the destructor of a unique pointer holding `ClientStatusReporting`
- `ClientStatusReporting` is then destroyed and it calls the destructor of `ClientStatusReportingNetwork`
- Then `ClientStatusReportingNetwork` is destructed, but, if I go with shared pointer (which `AccountPtr` stands for), the `Account` will never get destructed and this will introduce a memory leak for: `Account`, `ClientStatusReporting`, `ClientStatusReportingDatabase` and `ClientStatusReportingNetwork`

Reply:
OK,
I get it